src/hardware/gus.cpp

```diff
@@ -1,924 +1,1418 @@
-static Bit8u GUSRam[1024*1024]; // 1024K of GUS Ram
-static Bit16u vol16bit[4096];
-static Bit32u pantable[16];
-
-class GUSChannels;
-static void CheckVoiceIrq(void);
-
-struct GFGus {
-	Bit8u gRegSelect;
-	Bit16u gRegData;
-	Bit32u gDramAddr;
-	Bit16u gCurChannel;
-
-	Bit8u DMAControl;
-	Bit16u dmaAddr;
-	Bit8u TimerControl;
-	Bit8u SampControl;
-	Bit8u mixControl;
-	Bit8u ActiveChannels;
-	Bit32u basefreq;
-
-	struct GusTimer {
-		Bit8u value;
-		bool reached;
-		bool raiseirq;
-		bool masked;
-		bool running;
-		float delay;
-	} timers[2];
-	Bit32u rate;
-	Bitu portbase;
-	Bit8u dma1;
-	Bit8u dma2;
-
-	Bit8u irq1;
-	Bit8u irq2;
-
-	bool irqenabled;
-	bool ChangeIRQDMA;
-	// IRQ status register values
-	Bit8u IRQStatus;
-	Bit32u ActiveMask;
-	Bit8u IRQChan;
-	Bit32u RampIRQ;
-	Bit32u WaveIRQ;
-} myGUS;
-
-Bitu DEBUG_EnableDebugger(void);
-
-static void GUS_DMA_Callback(DmaChannel * chan,DMAEvent event);
-
-class GUSChannels {
+
+#define LOG_GUS 0 // set to 1 for detailed logging
+
+// Constants used in member definitions
+constexpr uint8_t ADLIB_CMD_DEFAULT = 85u;
+constexpr float AUDIO_SAMPLE_MAX = static_cast<float>(MAX_AUDIO);
+constexpr float AUDIO_SAMPLE_MIN = static_cast<float>(MIN_AUDIO);
+constexpr int BUFFER_FRAMES = 48;
+constexpr int BUFFER_SAMPLES = BUFFER_FRAMES * 2; // 2 samples/frame (left & right)
+constexpr uint32_t BYTES_PER_DMA_XFER = 8 * 1024; // 8 KB per transfer
+constexpr auto DELTA_DB = 0.002709201;             // 0.0235 dB increments
+constexpr uint8_t DMA_IRQ_ADDRESSES = 8u; // number of IRQ and DMA channels
+constexpr uint32_t ISA_BUS_THROUGHPUT = 32 * 1024 * 1024; // 32 MB/s
+constexpr uint16_t DMA_TRANSFERS_PER_S = ISA_BUS_THROUGHPUT / BYTES_PER_DMA_XFER;
+constexpr uint8_t MAX_VOICES = 32u;
+constexpr uint8_t MIN_VOICES = 14u;
+constexpr uint8_t MAX_DMA_ADDRESS = 7u;
+constexpr uint8_t MAX_IRQ_ADDRESS = 15u;
+constexpr uint8_t MIN_DMA_ADDRESS = 1u;
+constexpr uint8_t MIN_IRQ_ADDRESS = 2u;
+constexpr float MS_PER_DMA_XFER = 1000.0f / DMA_TRANSFERS_PER_S;
+constexpr float ONE_AMP = 1.0f; // first amplitude value
+constexpr uint8_t PAN_DEFAULT_POSITION = 7u;
+constexpr uint8_t PAN_POSITIONS = 16u;  // 0: -45-deg, 7: centre, 15: +45-deg
+constexpr uint32_t RAM_SIZE = 1048576u; // 1 MB
+constexpr uint8_t READ_HANDLERS = 8u;
+constexpr float SOFT_LIMIT_RELEASE_INC = AUDIO_SAMPLE_MAX * static_cast<float>(DELTA_DB);
+constexpr float TIMER_1_DEFAULT_DELAY = 0.080f;
+constexpr float TIMER_2_DEFAULT_DELAY = 0.320f;
+constexpr uint8_t VOICE_DEFAULT_STATE = 3u;
+constexpr int16_t VOLUME_INC_SCALAR = 512; // Volume index increment scalar
+constexpr auto VOLUME_LEVEL_DIVISOR = 1.0 + DELTA_DB;
+constexpr uint16_t VOLUME_LEVELS = 4096u;
+constexpr int16_t WAVE_WIDTH = 1 << 9; // Wave interpolation width (9 bits)
+constexpr float WAVE_WIDTH_INV = 1.0f / WAVE_WIDTH;
+constexpr uint32_t WAVE_MSW_MASK = (1 << 16) - 1; // Upper wave mask
+constexpr uint32_t WAVE_LSW_MASK = 0xffffffff ^ WAVE_MSW_MASK; // Lower wave mask
```

Comment:
Isn't this the same as `constexpr uint32_t WAVE_LSW_MASK = ~WAVE_MSW_MASK;`?

Reply:
I'm not sure; I'm using almost all of the masks / hex values as-is. Happy if you want to suggest/recommend any improvements or different ways to represent it.

Reply:
It's the same - https://onlinegdb.com/BJ9AlJ5zD; using your version instead.